org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/ui/preferences/XMLCatalogPreferencePage.java

```diff
@@ -0,0 +1,279 @@
+				entries.add(newEntry);
+				viewer.refresh();
+				viewer.setSelection(new TreeSelection(new TreePath(new String[] {newEntry})));
+			}
+		});
+		final Button changeButton = new Button(buttonComposite, SWT.PUSH);
+		changeButton.setLayoutData(new GridData(GridData.FILL_HORIZONTAL));
+		changeButton.setText(Messages.PreferencesPage_Change);
+		changeButton.setEnabled(false);
+		changeButton.addSelectionListener(new SelectionListener() {
+
+			public void widgetSelected(SelectionEvent e) {
+				if (selectedEntry == null) {
+					return;
+				}
+
+				String result = openSelectFileDialog(selectedEntry);
+				if (result == null || result.trim().length() == 0) {
+					return;
+				}
+				String newEntry = new File(result).getAbsolutePath();
+				entries.remove(selectedEntry);
+				entries.add(newEntry);
+				viewer.refresh();
+				viewer.setSelection(new TreeSelection(new TreePath(new String[] {newEntry})));
+
+			}
+
+			public void widgetDefaultSelected(SelectionEvent e) {
+
+			}
+		});
+		final Button removeButton = new Button(buttonComposite, SWT.PUSH);
+		removeButton.setLayoutData(new GridData(GridData.FILL_HORIZONTAL));
+		removeButton.setText(Messages.PreferencesPage_Remove);
+		removeButton.setEnabled(false);
+
+		removeButton.addSelectionListener(new SelectionListener() {
+
+			public void widgetSelected(SelectionEvent e) {
+				if (selectedEntry != null) {
+					entries.remove(selectedEntry);
+					viewer.refresh();
+				}
+			}
+
+			public void widgetDefaultSelected(SelectionEvent e) {
+			}
+		});
+
+		viewer.addSelectionChangedListener(new ISelectionChangedListener() {
+
+			public void selectionChanged(SelectionChangedEvent event) {
+				changeButton.setEnabled(false);
+				removeButton.setEnabled(false);
+				String oldSelectedEntry = selectedEntry;
+				selectedEntry = null;
+				ISelection selection = event.getSelection();
+				if (selection instanceof ITreeSelection) {
+					ITreeSelection treeSelection = (ITreeSelection) selection;
+					Object object = treeSelection.getFirstElement();
+					if (object instanceof String) {
+						selectedEntry = (String) object;
+						if (oldSelectedEntry != null && oldSelectedEntry.equals(selectedEntry)) {
+							selectedEntry = null;
+							viewer.setSelection(new TreeSelection(new TreePath(new String[] {})));
+						} else {
+							changeButton.setEnabled(true);
+							removeButton.setEnabled(true);
+						}
+					}
+				}
+			}
+		});
+
+		return composite;
+	}
+
+	@Override
+	public boolean performOk() {
+		String catalogsStr = "";
+		if (!entries.isEmpty()) {
+			for (String s : entries) {
+				catalogsStr += s + ",";
+			}
+		}
+		store.setValue(XMLPreferenceInitializer.XML_PREFERENCES_CATAGLOGS, catalogsStr);
+		return super.performOk();
+	}
+	
+	@Override
+	protected IPreferenceStore doGetPreferenceStore() {
+		return Activator.getDefault().getPreferenceStore();
+	}
+
+	private String openSelectFileDialog(String selectedFileName) {
```

Comment:
It would be a bit nicer if this could return a File.

Reply:
What should we do with file then?

Reply:
The think is that when receiving a String, we're not sure whether it represents the path (relative or absolute)?, the URI, the name... So passing a File clarifies everything, and later, consumers can reduce it to what's needed in their context.
Stronger typing FTW ;)

Reply:
I'm using `org.eclipse.swt.widgets.FileDialog` and its `open()` method returns an absolute file path as a `String`. So, in order to make it a `File` I will have to create a `File` for that `String`-value returned and then, later, convert it back to `String` when I need to show it in preference page or save it into the preference store.
Is such a double conversion is really needed?